src/compiler/checker.ts

```diff
@@ -9613,6 +9614,9 @@ namespace ts {
                             context.inferredTypes[index] = inferredType = instantiatedConstraint;
                         }
                     }
+                    if (context.useAnyForNoInferences && !inferences.length && inferredType === emptyObjectType) {

```

Comment:
Why this logic when we're already doing similar checks above? I would just modify the `inferredType = emptyObjectType` assignment above to `inferredType = context.usAnyForNoInferences ? anyType : emptyObjectType`.